Doc rework (development guide, part 1)

stephenfin opened the pull request:
This is combines two things:

- A rewrite of the existing tutorials to present a consistent style
- The addition of a new tutorial, based on a guide by @ofosos, which covers
  domains and cross-references

stephenfin committed:
docs: Rework the "helloworld" tutorial

Nothing to drastic going on here, but this format works better for other
tutorials we're adding.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
docs: Rework "todo" tutorial

Adopt the same format as was recently added in the "hello world"
tutorial, making this more of a walkthrough as expected from tutorials.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
doc: Use 'literalinclude' directive for examples

This avoid duplication and could conceivably let us test this stuff in
code later on.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
doc: Add "recipe" tutorial

This is based on a post from opensource.com [1] and demonstrates how one
can use indexes for cross-referencing and domains to group these indexes
along with domains and roles. The source code was taken from [2] after
getting the license changed [3].

[1] https://opensource.com/article/18/11/building-custom-workflows-sphinx
[2] https://github.com/ofosos/sphinxrecipes
[3] https://github.com/ofosos/sphinxrecipes/issues/1

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented on doc/development/tutorials/helloworld.rst:
```diff
@@ -1,39 +1,54 @@
-The objective of this tutorial is to create a very basic extension that adds a new
-directive that outputs a paragraph containing `hello world`.
+The objective of this tutorial is to create a very basic extension that adds a
+new directive. This directive will output a paragraph containing `hello world`.
```
I noticed single backtick is not good for this case. How about replacing by double-quotes?

stephenfin replied:
Can do.

tk0miya commented on doc/development/tutorials/helloworld.rst:
```diff
@@ -1,39 +1,54 @@
-#. Create an :file:`_ext` folder in :file:`source`.
+We want the extension to add the following to Sphinx:
+
+* A ``helloworld`` directive, that will simply output the text `hello world`.
```
This is also.

tk0miya commented on doc/development/tutorials/helloworld.rst:
```diff
@@ -56,30 +72,33 @@ Open :file:`helloworld.py` and paste the following code in it:
 
-   :doc:`todo`
+This class extends the docutils_' ``Directive`` class. All extensions that
+create directives should extend this class.
```
I wonder we should introduce `SphinxDirective` class here instead of docutils'.
It provides helper methods for Sphinx framework. For example, user directives can access the config object via `self.config` attribute. I believe it is very useful for developers. 

On the other hand, the directive becomes only for Sphinx. It can't use with other docutils declarative tools.

stephenfin replied:
I don't understand the difference, to be honest. In fact, I thought we had removed that in favour of the docutils `Directive` class, but maybe I'm thinking of something else. Do you see any reason to do this here or in a later tutorial?

tk0miya replied:
Hmm? removed? Now we're using `SphinxDirective` class to make custom directives.
https://github.com/sphinx-doc/sphinx/blob/6bdac80b5fb75641f7cd6f4b8c47bf6241360e2c/sphinx/directives/__init__.py#L49

I believe `SphinxDirective` is useful to develop Sphinx extension. So it would be better to introduce it in this tutorial.

stephenfin replied:
As noted below, I'm still not sure why to use this. Could you elaborate on why it's useful. I can add a note saying we're extending from the docutils directive class but one should extend from the Sphinx one if they need access to `self.config` or the likes?

tk0miya replied:
With docutils' Directive class, you need to write a complex code to access Sphinx objects (ex. `app`, `env`, `config`, `builder`).
For example, in case of a directive would like to access `myext_config` confval, you need to write following code:
```
class MyDirective(docutils.parsers.rst.Directive):
    def run(self):
        if self.state.document.settings.env.config.myext_config:
            ...
```

It is hard to write and understand for all developers.

With `SphinxDirective` class, you can access it simply with `self.config`:
```
class MyDirective(sphinx.util.docutils.SphinxDirective):
    def run(self):
        if self.config.myext_config:
            ...
```

If custom directives does not access to Sphinx objects, our `SphinxDirective` is not needed. So helloworld and todo extensions would not need to inherit it. But developers would need to know the way to access to Sphinx objects in real case. So I think here is good point to let them know about `SphinxDirective`.

stephenfin replied:
I've added this to the todo case. Let me know if you think this is still unclear.

tk0miya commented on doc/development/tutorials/helloworld.rst:
```diff
@@ -89,74 +108,93 @@ The ``nodes.paragraph`` class creates a new paragraph node. A paragraph
-   The first argument is the name of the directive itself as used in an rST file.
+The simplest thing you can do it call the :meth:`~Sphinx.add_directive`
+method, which is what we've done here. For this particular call, the first
+argument is the name of the directive itself as used in an rST file. In this
```
We usually use "reST" for abbreviation of reStructuredText.

tk0miya commented on doc/development/tutorials/todo.rst:
```diff
@@ -217,12 +327,17 @@ The node structure that the directive returns looks like this::
-------------------
+Event handlers are one of Sphinx's most powerful features, providing a way to do
+hook into any part of the documentation process. There are many hooks available,
+as detailed in :doc:`/extdev/appapi`, and we're going to use a subset of them
```
I think using `:ref:` is better to here. `/extdev/appapi` is a page for Application APIs, not only for event handlers.

stephenfin replied:
Can do

tk0miya commented on doc/development/tutorials/todo.rst:
```diff
@@ -283,17 +398,149 @@ to be done::
+items are composed of the nodes from the ``todo`` entry and docutils nodes
+created on the fly: a paragraph for each entry, containing text that gives the
+location, and a link (reference node containing an italic node) with the
+backreference.  The reference URI is built by ``app.builder.get_relative_uri``
```
We already have document for `get_relative_uri()`. So it's better to refer it here.

tk0miya commented on doc/development/tutorials/todo.rst:
```diff
@@ -283,17 +398,149 @@ to be done::
 
-It is a bit more involved.  If our new "todo_include_todos" config value is
-false, all todo and todolist nodes are removed from the documents.
+The :event:`doctree-resolved` event is emitted at the end of :ref:`phase 3
```
I feel it is better to use name of phases here instead. It is related with the event name.

tk0miya commented on doc/development/tutorials/todo.rst:
```diff
@@ -283,17 +398,149 @@ to be done::
+* :meth:`~Sphinx.add_node` adds a new *node class* to the build system.  It also
+  can specify visitor functions for each supported output format.  These visitor
+  functions are needed when the new nodes stay until :ref:`phase 4 <build-phases>`
+  -- since the ``todolist`` node is always replaced in :ref:`phase 3 <build-phases>`,
```
Here is also. Let's replace phases by their names.

tk0miya commented on doc/development/tutorials/todo.rst:
```diff
@@ -283,17 +398,149 @@ to be done::
+
+   sys.path.append(os.path.abspath("./_ext"))
+
+   extensions = ['helloworld']
```
`todo`?

tk0miya commented on doc/development/tutorials/helloworld.rst:
```diff
@@ -115,10 +99,10 @@ the ``text`` parameter.
+.. literalinclude:: examples/helloworld.py
+   :language: python
+   :linenos:
+   :lines: 12-
```
How about using `:pyobject: setup` option instead? The line numbers are very fragile. So the option makes our document robust.

tk0miya replied:
ah, it seems other inclusions contain multiple python objects. So `:pyobject:` would not help us in this case...

stephenfin replied:
Yeah. I don't think this will we be too much hassle, in any case. If it would help, I could put a README in the folder saying that people should update the line numbers in the docs if they change the file.

stephenfin replied:
Added `README.rst`

tk0miya replied:
It seems not added. But I don't think the file will help us...

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+            name = '{}.{}.{}'.format('rcp', type(self).__name__, sig)
+            imap = self.env.domaindata['rcp']['obj2ingredient']
+            imap[name] = list(self.options.get('contains').split(' '))
+            objs = self.env.domaindata['rcp']['objects']
```
I think accessing internal data of domain directly is not good way. How about using domain methods like following:
```
recipes = self.env.get_domain('rcp')
recipes.add_recipe(...)
```

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+class IngredientIndex(Index):
+    """A custom directive that creates an ingredient matrix."""
+
+    name = 'ing'
```
Why are you using abbreviation here?

stephenfin replied:
I just copy-pasted this from the blog (see the commit messages). This is probably a question for @ofosos

stephenfin replied:
Changed to `ingredient`

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+        super(IngredientIndex, self).__init__(*args, **kwargs)
+
+    def generate(self, docnames=None):
+        """Return entries for the index given by *name*.
```
Is this docstring needed? It would be better to move this to the document.

Note: `RecipeIndex.generate()` also has same docstring. It is duplicated and hard to read the example code.

stephenfin replied:
Yes, I'll move it

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+        for name, ingr in ingr.items():
+            for ig in ingr:
+                imap.setdefault(ig,[])
+                imap[ig].append(name)
```
It would be better to add API to domain class.

stephenfin replied:
Could you give me an example? I'm not sure what you mean. Do you mean adding a `add_recipe` method?

stephenfin replied:
I guessed you meant `add_recipe` and did this

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+
+class RecipeDomain(Domain):
+
+    name = 'rcp'
```
Why abbreviation too?

stephenfin replied:
`rcp` is short for `recipe`, I guess. What would you suggest?

stephenfin replied:
Used `recipe` as noted elsewhere.

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+    StandardDomain.initial_data['anonlabels']['recipeindex'] = (
+        'rcp-rcp', '')
+    StandardDomain.initial_data['anonlabels']['ingredientindex'] = (
+        'rcp-ing', '')
```
I feel touching `StandardDomain` is very hacky solution. So it must not be introduced as an example.

stephenfin replied:
OK. How should we do this?

tk0miya replied:
Unfortunately, no alternative APis.

stephenfin replied:
@ofosos What was the reason for doing this? I can remove it and things work as expected. Is it necessary?

ofosos replied:
I think the reason for this was to have the indices show up in the TOC.

tk0miya replied:
Okay, I understood. But there are no way to add cross-reference target from extensions. If you've still needed it, please file a new issue. I'll consider to add new API.

stephenfin replied:
In the interim, I've simply removed this code. We can think about this long term.

tk0miya commented on doc/development/tutorials/helloworld.rst:
```diff
@@ -89,74 +108,93 @@ The ``nodes.paragraph`` class creates a new paragraph node. A paragraph
+.. code-block:: python
+
+   def setup(app):
+       app.add_directive("helloworld", HelloWorld)
```
I'd forgotten to mention the return value of `setup()`.
It would be better to metadata of the extension.
https://www.sphinx-doc.org/en/master/extdev/index.html#extension-metadata

stephenfin replied:
Good point. I'll do this for the other ones too

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+We want the extension to add the following to Sphinx:
+
+* A ``recipe`` :term:`directive`, containing some content describing the recipe
+  steps, along with a ``:contains:`` argument highlighting the main ingredients
```
`:contains:` is a directive option, not an argument. Arguments of directive are parameters following `::` mark:
```
.. directive-name::  this is arguments for directive
   :directive-option1: parameter for the option
   :directive-option2: parameter

   contents of directives.
   blah blah blah...
```

stephenfin replied:
Ah, I was using the argparse terminology of "positional arguments" and "optional arguments". My mistake

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+  of the recipe.
+
+* A ``reref`` :term:`role`, which provides a cross-reference to the recipe
+  itself.
```
I feel `ref` is better for such pupose. Users can use it via `:rcp:ref:`. (This is why I prefer to use non abbreviated domain name. Then we can use `:recipe:ref:` instead.)

stephenfin replied:
Good call. I'll change this.

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+
+      └── source
+          ├── _ext
+          │   └── todo.py
```
`recipe.py`?

stephenfin replied:
Whoops

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+  ``reref`` role
+
+
+Prerequisites
```
I wonder this section is really needed for each example. It is too much duplicated. I understand verbosity helps newbies. But hyperlinks can also help them.

stephenfin replied:
Yes, I wanted to avoid telling people to go back and forth between documents. Let's see if I can make it shorter though.

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+
+.. rubric:: The directive class
+
+The first thing to examine is the ``RecipeNode`` directive:
```
`RecipeDirective` class?

stephenfin replied:
Whoops

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+We also see that this directive defines ``has_content``, ``required_arguments``
+and ``option_spec``. Unlike the ``TodoDirective`` directive added in the
+:doc:`previous tutorial <todo>`, this directive takes a single argument, the
+recipe name, and an optional argument, ``contains``, in addition to the nested
```
`contains` is an option, not "argument".

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+
+Both ``IngredientIndex`` and ``RecipeIndex`` are derived from :class:`Index`.
+They implement custom logic to generate a tuple of values that define the
+index. Note that ``RecipeIndex`` is a degenerate index that has only one entry.
```
What "degenerate index" means?

stephenfin replied:
I'm not sure. This was copy-pasted. @ofosos?

ofosos replied:
It's just an index with a single 'category', meaning there are not multiple index categories that then have entries.

ofosos replied:
I'm not sure, probably you want the wording changed to say 'simple' or such. In retrospect that wording was not brilliant :)

tk0miya replied:
Simply, I'm not good at English :-) It's okay if you feel natural.

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+``objects`` and ``obj2ingredient``. These contain a list of all objects defined
+(i.e. all recipes) and a hash that maps a canonical ingredient name to the list
+of objects. The way we name objects is common across our extension and is
+defined in the ``get_full_qualified_name`` method. For each object created, the
```
Oh, `Domain.get_full_qualified_name()` is undocumented... we have to add document for it.

tk0miya replied:
My comment above is wrong. It is already documented (but it says nothing...)
https://www.sphinx-doc.org/en/master/extdev/domainapi.html

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+its target name. We'll use ``target`` to resolve our destination from our
+domain's ``objects`` because we currently have only one type of node.
+
+Moving on, we can see that we've defined two items in ``intitial_data``:
```
Note: `initial_data` will be copied to the `env.domaindata[domain_name]` as an initial data of the domain. And domain instance can access it via `self.data`.

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    initial_data = {
+        'objects': [],  # object list
```
Nobody restricts the name of keys. So it would be easy to understand if we rename this to `recipes` key.

stephenfin replied:
+1

tk0miya commented on doc/development/tutorials/recipe.rst:
```diff
@@ -0,0 +1,224 @@
+(i.e. all recipes) and a hash that maps a canonical ingredient name to the list
+of objects. The way we name objects is common across our extension and is
+defined in the ``get_full_qualified_name`` method. For each object created, the
+canonical name is ``rcp.<typename>.<objectname>``, where ``<typename>`` is the
```
I feel `<typename>` is not needed in this case. What about `recipe.<recipe-name>` instead? (`<recipe-name>` is same as object name).

The full qualified name is used to look up the objects from indices in intersphinx extension. So it is okay if enough unique. The intersphinx extension searches the object from the return value of `get_objects()`. So The form of the qualified should be used in the method.

stephenfin replied:
+1

stephenfin committed:
docs: Address review comments

helloworld:
- Return version metadata from extension
- Use 'reST' instead of 'rST'
- Don't use single backticks

todo:
- Return version metadata from extension
- Link to events section of API guide, rather than entire document
- Include name of phases in when describing build phases
- Use more method cross-references where possible
- Fix typo in usage example

recipe:
- Return version metadata from extension
- Rework code to simplify things
- Remove docstrings from 'generate' functions, which are simply
  duplicates of the original docstring
- Rename 'rcp' directive to 'recipe', the 'reref' role to 'ref', and a
  whole lot of variables to something more grokable
- Fix typos in both documentation and code

I've also fixed up the docstrings for some of the 'domain' functions to
make them render a little nicer (they took me a while to figure out).

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
@tk0miya Think I've addressed all but one comment. The only thing let is using the `SphinxDirective` instead of `Directive`. Could you elaborate what this does so I know why I should use it? I'll include your explanation in the docs.

Also, I'm thinking it would be good to actually test these examples somehow. Have you suggestions for how to do this, @tk0miya?

tk0miya commented:
> Also, I'm thinking it would be good to actually test these examples somehow. Have you suggestions for how to do this, @tk0miya?

We've provided `sphinx.testing` to test Sphinx itself and extensions. So it would be nice if we'll test our examples. I suppose that would also be a kind of tutorial which developers would like to know.

tk0miya commented on doc/development/tutorials/examples/recipe.py:
```diff
@@ -0,0 +1,171 @@
+    shortname = 'Ingredient'
+
+    def __init__(self, *args, **kwargs):
+        super(IngredientIndex, self).__init__(*args, **kwargs)
```
It seems this method does not anything. How about removing?

stephenfin replied:
Good call. Done.

stephenfin committed:
docs: Address further review comments

todo:
- Subclass SphinxDirective instead of Directive

recipe:
- Remove unnecessary '__init__' methods

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
> > Also, I'm thinking it would be good to actually test these examples somehow. Have you suggestions for how to do this, @tk0miya?
> 
> We've provided `sphinx.testing` to test Sphinx itself and extensions. So it would be nice if we'll test our examples. I suppose that would also be a kind of tutorial which developers would like to know.

OK, let's do this in a follow-up.

stephenfin commented:
@tk0miya anything else needed here?

tk0miya commented:
@stephenfin It seems flake8 emits warnings to example codes. Could you check them please?
https://travis-ci.org/sphinx-doc/sphinx/jobs/494899313#L232-L238

stephenfin committed:
docs: Remove unused imports

These were causing flake8 failures.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin commented:
> @stephenfin It seems flake8 emits warnings to example codes. Could you check them please?
> [travis-ci.org/sphinx-doc/sphinx/jobs/494899313#L232-L238](https://travis-ci.org/sphinx-doc/sphinx/jobs/494899313#L232-L238)

That's what I get for posting from my phone. I should have spotted that. Fixed now.

tk0miya commented:
Just merged!
